### ISSUE TITLE
PKINIT AS-REQ with name-type of NT-X500-PRINCIPAL cannot determine principal from subject DN

### ISSUE THREAD

**Issue by svmhdvn** [heimdal/heimdal#1099]
**Describe the bug**
In a usual unix testing setup with a heimdal KDC and a heimdal-bundled  `kinit` client, PKINIT PKCS11 authentication uses a `name-type` of `NT-PRINCIPAL` and a `CNameString` of the principal name without the realm. However, on windows, interactive kerberos smartcard (PKINIT) authentication sends an AS-REQ to the KDC with a `name-type` of `NT-X500-PRINCIPAL` and a `CNameString` of the subject DN extracted from the x509 certificate. Heimdal reports an error of `UNKNOWN -- CN=principal@REALM: no such entry found in hdb`.

As observed, the entire string `CN=principal` is being treated like a principal name, whereas it should extract the principal name from the CN. Otherwise, perhaps another type such as `uid=principal` should be used.

My understanding is that this means Heimdal cannot extract the principal name from the common name (CN) found in the subject DN of the x509 certificate. There also doesn't seem to be any config options that allow subject DN matching.

I've also tried using `--pkinit-acl` and `--alias` options with `kadmin modify` on the principals, but none of them have any effect. Same with a `/var/heimdal/pki-mapping` file.

krb5.conf:
```
[...]
[realms]
        REALM = {
                pkinit_require_eku = true
                pkinit_require_krbtgt_otherName = true
        }
[kdc]
        enable-pkinit = yes
        pkinit_principal_in_certificate = yes
[...]
```

**To Reproduce**
1. Setup a simple heimdal KDC with PKINIT enabled with the above krb5.conf additional config options
2. Use the info manual on setting up PKINIT to correctly create x509 certs for the CA, KDC, and client (principal) certs
3. Import the client key and cert onto a smartcard (I used a yubikey 5)
4. Setup a simple samba CIFS share with kerberos enabled (correctly setup service principal and keytab installed). You can also use any other method of spawning an interactive smartcard login with kerberos on windows.
5. Make an authenticated request to spawn a login prompt
6. Watch the kerberos verbose logs and observe previously stated error message 

**Expected behavior**
Both heimdal `kinit` and windows smartcard kerberos login should succeed with the same smartcard x509 cert.

**Screenshots**
Will attach screenshot of windows smartcard login prompt soon.

**Additional context**
KDC: heimdal 7.8.0 running in FreeBSD 13.2
Samba 4.16 used in FreeBSD 13.2 with keytab from KDC
Windows 11 client connecting to Samba CIFS share

CIFS authentication is successful when using a password based regular kerberos login with no PKINIT.

Let me know if you need any additional info.

**Comment by nicowilliams** [comment]
> However, on windows, interactive kerberos smartcard (PKINIT) authentication sends an AS-REQ to the KDC with a name-type of `NT-X500-PRINCIPAL` and a `CNameString` of the subject DN extracted from the x509 certificate.

Yes, Heimdal currently lacks support for `KRB5_NT_X500_PRINCIPAL`.  It probably wouldn't be hard to add.
